quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -172,13 +188,29 @@ class Monad m => RunModel state m where
   postcondition :: forall a. (state, state) -> Action state a -> LookUp m -> Realized m a -> PostconditionM m Bool
   postcondition _ _ _ _ = pure True
 
+  -- | Postcondition on the `a` value produced at some _negative_ step.
+  -- The result is `assert`ed and will make the property fail should it be `False`. This is useful
+  -- to check the implementation produces expected errors.
+  negativePostcondition :: forall a. (state, state) -> Action state a -> LookUp m -> Realized m a -> PostconditionM m Bool
```

Comment:
Same remark as  `negativePrecondition`: `negativePostcondition` does not really convey the semantics of this function. `actionFailure` perhaps? 
Perhaps what happens here is that an implementation detail (the `Polarity`) leaks in the API?

Reply:
I'm not sure what you mean by it leaking from the API? The issue is that you wany to write your code with happy path in mind and when you're adding negative tests you don't want to perform the polarity check in the existing postcondition but instead write a new condition for e.g. what HTTP error codes you get back from negative actions.

One could imagine a different name for the function, but I'm not sure what name best conveys the semantics that this really is a postconditon on the failure semantics.

Reply:
"leaking" in a metaphorical sense: I don't think in terms of "polarity" when it comes to writing a model defining the behaviour I expect, this seems to be a way to implement some feature of the model.
 
I don't think your example conveys the right idea about what negative testing is in the context. Testing that some HTTP call returns some error code is not a negative test in my opinion, it's just a plain test testing some behaviour of the SUT in case of failures.

That's perhaps the part where I am unsure about what this feature is about. If I think of q-d model-based testing in terms of state machines or automatas, I imagine "negative testing" to be a simple and generic way to _complete_ the transition graph of the model such that from every state, every action that's not explicitly allowed (eg. whose `precondition` is false) should either leave the state unchanged, or should lead to a sink state from which no other action can be taken.

> I'm not sure what name best conveys the semantics that this really is a postconditon on the failure semantics.

`postconditionOnFailure` would be a better name IMO.

Reply:
Perhaps explaining in the `StateModel`'s documentation what negative testing is in the context of q-d would help?

Reply:
That is indeed the goal - and what you've just written is perfectly within reach of this extension. In my view that's what this is doing. If you look atbthe registry example we explicitly don't define the negative state transition - so they should be `id`. However, I think explaining this in the terms you've given above and changing the names might go a long way towards addressing your concerns?

Reply:
Great, it makes sense. I think yes, having some explanation somewhere might help. Happy to contribute that later, I am off for 10 days for now but might find some time to write some doc and perhaps use Hydra again as an example.

Reply:
Perfect! I'll try to write something up in the comments of these files and make it more clear (with some documentation) in the registry example when I have time to return to this too!

I'll be off soon too so no sweat here.